File: src/main/java/algo/backtrack/CombinationSum_III_216.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.List;

public class CombinationSum_III_216 {
    public static List<List<Integer>> combinationSum3(int k, int n) {
        var combinations = new ArrayList<List<Integer>>();
        backtrack(k, n, 1, new ArrayList<>(), combinations);
        return combinations;
    }

    private static void backtrack(int k, int n, int start, List<Integer> combination, List<List<Integer>> all) {
        if (k == 0 && n == 0) {
            all.add(new ArrayList<>(combination));
            return;
        }
        for (var i = start; i <= 9; i++) {
            if (n < i) {
                break;
            }
            combination.add(i);
            backtrack(k - 1, n - i, i + 1, combination, all);
            combination.removeLast();
        }
    }

    public static void main(String[] args) {
        combinationSum3(3, 7).forEach(System.out::println);
    }
}

File: src/main/kotlin/algo/backtrack/CombinationSum_III_216.kt
package algo.backtrack

class KtCombinationSum_III_216 {
    fun combinationSum3(k: Int, n: Int): List<List<Int>> {
        val all: MutableList<List<Int>> = mutableListOf()
        fun backtrack(k: Int, sum: Int, start: Int, list: MutableList<Int>) {
            if (sum < 0) {
                return
            }
            if (k == 0 && sum == 0) {
                all.add(list.toList())
                return
            }
            for (num in start..9) {
                val remain = sum - num
                if (remain >= 0) {
                    list.add(num)
                    backtrack(k - 1, remain, num + 1, list)
                    list.removeLast()
                }
            }
        }
        backtrack(k, n, 1, mutableListOf())
        return all
    }
}

fun main() {
    val sum3 = KtCombinationSum_III_216()
    println(sum3.combinationSum3(3, 7))
    println(sum3.combinationSum3(3, 9))
}